### PR TITLE
[Merged by Bors] -  feat: add commutative diagram widget

### PR DESCRIPTION
**Pull request by Vtec234**
Adds a widget for displaying goals in the language of category theory as commutative diagrams. Examples are found in `test/CommDiag.lean`. The set of diagram shapes which can be visualized is extensible via the `@[expr_presenter]` attribute. The widget is triggered in a tactic proof using the `with_panel_widgets [GoalTypePanel]` tactic combinator. Support for globally enabling the widget in a given file is planned for another PR.

Continued from #363, please see there for more discussion.

### CODE REVIEW

File: Mathlib/Tactic/Widget/CommDiag.lean [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+@[expr_presenter]
+def commutativeSquarePresenter : ExprPresenter where
+  userName := "Commutative square"
+  layoutKind := .block
+  present type := do
+    if let some d ← commSquareM? type then
+      return d
+    throwError "Not a commutative square."
+
+/-! ## Example diagrams -/
```
Comment (semorrison) [inline Mathlib/Tactic/Widget/CommDiag.lean:139]: Perhaps move these to `test/CommDiag.lean`?

**Comment by semorrison** [comment]
This looks promising to me, and works out of the box.

From a UI perspective:
* even when I click into examples, it took me a while to work out I needed to click on the word "default" to find the other display options. More aggressively showing the options, or indeed more aggressively just rendering a non-default option, would be preferable.
* `withPanelWidgets [SelectionPanel]` is too "low-level" an incantation for someone trying to prove an actual theorem about categories to use in practice. How could be make it more user friendly to invoke?

**Comment by semorrison** [comment]
@Vtec234, should this have the `WIP` label still? Do you want to do anything about my comments above (not necessary; up to you).

**Comment by Vtec234** [comment]
> From a UI perspective: ..

These are both good questions, and they have not been figured out in this PR! Luckily they are orthogonal to its core, which is just registering two `Expr` presenters for squares and triangles. Once the registration is done, the output of the presenter can be displayed by any kind of top-level widget, including the selection one, but also any alternative such as a 'just show the target type thanks' widget, or anything else. I.e., the shift-click/dropdown mechanism (which I agree is awkward) is not part of the PR, other than being used in the test code.

> How could be make it more user friendly to invoke?

Potentially EdAyers/ProofWidgets4#5 may help. The idea would be to put `show_panel_widgets [CommDiagWidget]` at the top of the file, and from then on it would show up in every proof.

**Comment by Vtec234** [comment]
I'm not sure about the `WIP` label. The prototype works in the test, but as noted it needs a better way to trigger it for use in actual proofs. Also we need more metaprograms to actually generate diagram descriptions (by extracting the objects/homs out of an identity type).  Additionally, the Penrose widget could use robustness improvement, but that is contained within ProofWidgets and would just be a version bump in mathlib.

I will not be able to get to any of these until mid-May at least, so it's up to mathlib whether it prefers to move quickly or wait for more features :)

**Comment by semorrison** [comment]
Sometimes the nodes displayed aren't big enough for the text:

Screen Shot 2023-05-03 at 2 05 35 pm

When selecting the same subexpression a second time it seems to work fine.

File: Mathlib/Tactic/Widget/CommDiag.lean [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+/-! ## Commutative triangles -/
+
+/--
+Triangle with `homs = [f,g,h]` and `objs = [A,B,C]`
+```
+A f B
+  h g
+    C
+``` -/
```
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/CommDiag.lean:77]: This seems to have broken the syntax higlighter. I think
````suggestion
/--
Triangle with `homs = [f,g,h]` and `objs = [A,B,C]`
```
A f B
  h g
    C
```
-/
````
fixes it

File: widget/src/penrose/commutativeOpt.sty
Comment (eric-wieser) [inline widget/src/penrose/commutativeOpt.sty:1]: Can we put a readme in this `penrose` folder explaining what language these files are in?

File: Mathlib/Tactic/Widget/CommDiag.lean [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+/-- Given a Hom type `α ⟶ β`, return `(α, β)`. Otherwise `none`. -/
+def homType? (e : Expr) : Option (Expr × Expr) := do
+  let some (_, _, A, B) := e.app4? ``Quiver.Hom | none
+  return (A, B)
+
+/-- Given composed homs `g ≫ h`, return `(g, h)`. Otherwise `none`. -/
+def homComp? (f : Expr) : Option (Expr × Expr) := do
+  let some (_, _, _, _, _, f, g) := f.app7? ``CategoryStruct.comp | none
+  return (f, g)
```
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/CommDiag.lean:49]: Does it make sense to use `Qq` here so that `Lean.Expr.app7?` isn't needed?

Comment (digama0) [reply Mathlib/Tactic/Widget/CommDiag.lean:49]: Qq match is in MetaM and uses defeq, which is a lot more costly.

**Comment by eric-wieser** [comment]
Can you update the PR description with something suitable for a final commit message?

**Comment by eric-wieser** [comment]
Can the colors be made to reflect the color scheme somehow? Having the arrows in the main text color would make it a lot easier to read:

image

File: Mathlib/Tactic/Widget/CommDiag.lean [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+      ("f", f), ("g", g), ("h", lhs)]
+
+@[expr_presenter]
+def commutativeTrianglePresenter : ExprPresenter where
+  userName := "Commutative triangle"
+  layoutKind := .block
+  present type := do
+    if let some d ← commTriangleM? type then
+      return d
+    throwError "Couldn't find a commutative triangle."
```
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/CommDiag.lean:106]: When I hit this error, I get
```
Error: Rpc error: InternalError: Couldn't find a commutative triangle.
```
Do we need the word "error" three times?

Comment (digama0) [reply Mathlib/Tactic/Widget/CommDiag.lean:106]: that's probably not something that we can do anything about without changing core. This looks like the default error repr printer, applied to something like `InternalError.mk (RpcError.mk e)`.

Comment (Vtec234) [reply Mathlib/Tactic/Widget/CommDiag.lean:106]: This is not the Lean core printer, it's something that happens in JS. I think for displaying errors it is better to err (heh) on the side of verbosity/more information. The real issue is that this error is displayed at all during normal operation.

File: widget/src/penrose/README.md [hunk 5]
```diff
@@ -0,0 +1,4 @@
+This directory contains source files for Penrose (https://penrose.cs.cmu.edu/) diagrams.
+Domain files (.dsl) describe the objects available in a given domain.
+Style files (.sty) specify how diagrams involving the objects from a given domain should be drawn.
+Substance files (.sub) describe specific diagrams.
```
Comment (eric-wieser) [inline widget/src/penrose/README.md:4]:
```suggestion
This directory contains source files for [Penrose diagrams]((https://penrose.cs.cmu.edu/):
* Domain files (`.dsl`) describe the objects available in a given domain.
* Style files (`.sty`) specify how diagrams involving the objects from a given domain should be drawn.
* Substance files (`.sub`) describe specific diagrams.
```

And maybe add a comment referring to `ProofWidgets4` supporting these.

**Comment by Vtec234** [comment]
Okay, it's ready! @semorrison / @eric-wieser could you check that it still works, please?

>  More aggressively showing the options, or indeed more aggressively just rendering a non-default option, would be preferable.

We always show a widget version over non-widget ones now, and a system for user-defined precedences between different widgets is planned.

> Sometimes the nodes displayed aren't big enough for the text:

I rewrote the rendering code to be much more robust. If it ever shows an inconsistent state, it will be during loading, and the widget is grayed out then.

> Can the colors be made to reflect the color scheme somehow?

They do now.

> The idea would be to put `show_panel_widgets [CommDiagWidget]` at the top of the file, and from then on it would show up in every proof.

This is the only outstanding item left. I'd like to do it in another PR as changes to core are required.

**Comment by semorrison** [comment]
Still works for me!

**Comment by semorrison** [comment]
bors merge

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**